### PR TITLE
Fix #3147: Support building on JDK 9.

### LINKED ISSUE #3147: Support building on JDK 9
I am working on a new Scala 2.12 community build on JDK 9 (https://github.com/scala/community-builds/issues/609)

Naturally, one of the first projects I would like to include is Scala.js

but I'm getting `loaded from class ExposedValues$ in package with name ExposedValues$ and classloader java.net.URLClassLoader@2f29400e` when loading the build definition

the underlying cause is https://github.com/scala/scala-dev/issues/304

one rather slow possible way forward here is to wait for Scala 2.12.4 to be released with the fix for 304, then wait for a sbt 1.0.x release with the upgrade to 2.12.4, then upgrade Scala.js from sbt 0.13.x to 1.0.x

a faster, but more annoying, way forward would be to figure out which sbt plugin that Scala.js uses has its code in an empty package, and fix that plugin. (there could be more than one)

I see there is a ticket on upgrading Scala.js's own build to sbt 1.0.x: https://github.com/scala-js/scala-js/issues/3069

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Locally tested on a JDK 9 with ++2.12.3. Note that Scala 2.11 simply crashes with `MissingRequirementError`, but at this point I don't think we can do anything about that one.

**Comment by SethTisue** [comment]
I tried to build this branch in the JDK 9 community build but I got:

```
[scala-js] java.lang.NullPointerException
[scala-js] 	at build.Build$$anonfun$24.apply(Build.scala:165)
[scala-js] 	at build.Build$$anonfun$24.apply(Build.scala:161)
```

the lines of code in question are:

```
            System.getProperty("sun.boot.class.path")
              .split(java.io.File.pathSeparator)
```

but:

```
Welcome to Scala 2.12.3 (Java HotSpot(TM) 64-Bit Server VM, Java 9).
Type in expressions for evaluation. Or try :help.

scala> System.getProperty("sun.boot.class.path")
res0: String = null
```

**Comment by sjrd** [comment]
Ah, that happens on `library/doc` (and `library/publishLocal`), which I had not tested, indeed.

Do you know if there is any way to get the same information on JDK 9? i.e., identify what is the `.jar` of the Java base library, so that sbt can instruct Scaladoc to the do the proper API wiring?

In the meantime, an easy workaround is to comment out [this `apiMappings +=` setting](https://github.com/scala-js/scala-js/blob/b8fc74515491880c7b54ff31b7afb7a8b7fa3311/project/Build.scala#L207-L221). Although then I get this other error on `library/doc`:
```
[error]   bad constant pool tag 20 at byte 108
[error]      while compiling: /localhome/doeraene/projects/scalajs/library/src/main/scala/scala/scalajs/reflect/annotation/EnableReflectiveInstantiation.scala
[error]         during phase: globalPhase=terminal, enteringPhase=parser
[error]      library version: version 2.12.3
[error]     compiler version: version 2.12.3
[error]   reconstructed args: -implicits -bootclasspath /home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar -feature -P:scalajs:sjsDefinedByDefault -P:scalajs:suppressExportDeprecations -unchecked -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar#http://www.scala-js.org/api/scalajs-library/0.6.21-SNAPSHOT/ -classpath /home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar -encoding utf8 -deprecation -groups -Xfatal-warnings -d /localhome/doeraene/projects/scalajs/library/target/scala-2.12/api -Xplugin:/localhome/doeraene/projects/scalajs/compiler/target/scala-2.12/scalajs-compiler_2.12.3-0.6.21-SNAPSHOT.jar
[error]
[error]   last tree to typer: Ident(<argument>)
[error]        tree position: <unknown>
[error]             tree tpe: scala.scalajs.reflect.InstantiatableClass
[error]               symbol: <none>
[error]    symbol definition: <none> (a NoSymbol)
[error]       symbol package: <none>
[error]        symbol owners:
[error]            call site: package <root> in <none>
[error]
[error] <Cannot read source file>
scala.reflect.internal.FatalError:
  bad constant pool tag 20 at byte 108
     while compiling: /localhome/doeraene/projects/scalajs/library/src/main/scala/scala/scalajs/reflect/annotation/EnableReflectiveInstantiation.scala
        during phase: globalPhase=terminal, enteringPhase=parser
     library version: version 2.12.3
    compiler version: version 2.12.3
  reconstructed args: -implicits -bootclasspath /home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar -feature -P:scalajs:sjsDefinedByDefault -P:scalajs:suppressExportDeprecations -unchecked -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar#http://www.scala-js.org/api/scalajs-library/0.6.21-SNAPSHOT/ -classpath /home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar -encoding utf8 -deprecation -groups -Xfatal-warnings -d /localhome/doeraene/projects/scalajs/library/target/scala-2.12/api -Xplugin:/localhome/doeraene/projects/scalajs/compiler/target/scala-2.12/scalajs-compiler_2.12.3-0.6.21-SNAPSHOT.jar

  last tree to typer: Ident(<argument>)
       tree position: <unknown>
            tree tpe: scala.scalajs.reflect.InstantiatableClass
              symbol: <none>
   symbol definition: <none> (a NoSymbol)
      symbol package: <none>
       symbol owners:
           call site: package <root> in <none>

<Cannot read source file>
        at scala.reflect.internal.Reporting.abort(Reporting.scala:61)
        at scala.reflect.internal.Reporting.abort$(Reporting.scala:57)
        at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:16)
        at scala.tools.nsc.symtab.classfile.ClassfileParser$ConstantPool.errorBadTag(ClassfileParser.scala:366)
        at scala.tools.nsc.symtab.classfile.ClassfileParser$ConstantPool.<init>(ClassfileParser.scala:191)
        at scala.tools.nsc.symtab.SymbolLoaders$ClassfileLoader$classfileParser$.newConstantPool(SymbolLoaders.scala:294)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.$anonfun$parse$1(ClassfileParser.scala:155)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.parse(ClassfileParser.scala:127)
        at scala.tools.nsc.symtab.SymbolLoaders$ClassfileLoader.doComplete(SymbolLoaders.scala:316)
        at scala.tools.nsc.symtab.SymbolLoaders$SymbolLoader.complete(SymbolLoaders.scala:218)
        at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1531)
        at scala.tools.nsc.doc.base.MemberLookupBase.$anonfun$lookupInTemplate$4(MemberLookupBase.scala:149)
        at scala.tools.nsc.doc.base.MemberLookupBase.$anonfun$lookupInTemplate$4$adapted(MemberLookupBase.scala:149)
        at scala.collection.TraversableLike.$anonfun$filterImpl$1(TraversableLike.scala:248)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike.filterImpl(TraversableLike.scala:247)
        at scala.collection.TraversableLike.filterImpl$(TraversableLike.scala:245)
        at scala.collection.AbstractTraversable.filterImpl(Traversable.scala:104)
        at scala.collection.TraversableLike.filter(TraversableLike.scala:259)
        at scala.collection.TraversableLike.filter$(TraversableLike.scala:259)
        at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
        at scala.tools.nsc.doc.base.MemberLookupBase.cleanupBogusClasses$1(MemberLookupBase.scala:149)
        at scala.tools.nsc.doc.base.MemberLookupBase.lookupInTemplate(MemberLookupBase.scala:160)
        at scala.tools.nsc.doc.base.MemberLookupBase.lookupInTemplate(MemberLookupBase.scala:124)
        at scala.tools.nsc.doc.base.MemberLookupBase.lookupInRootPackage(MemberLookupBase.scala:111)
        at scala.tools.nsc.doc.base.MemberLookupBase.memberLookup(MemberLookupBase.scala:52)
        at scala.tools.nsc.doc.base.MemberLookupBase.memberLookup$(MemberLookupBase.scala:48)
        at scala.tools.nsc.doc.DocFactory$$anon$1.memberLookup(DocFactory.scala:68)
        at scala.tools.nsc.doc.base.MemberLookupBase$$anon$1.link$lzycompute(MemberLookupBase.scala:27)
        at scala.tools.nsc.doc.base.MemberLookupBase$$anon$1.link(MemberLookupBase.scala:27)
        at scala.tools.nsc.doc.base.comment.EntityLink$.unapply(Body.scala:74)
        at scala.tools.nsc.doc.html.Page$.inlineToStr(Page.scala:29)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1(Page.scala:18)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1$adapted(Page.scala:18)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
        at scala.collection.immutable.List.flatMap(List.scala:352)
        at scala.tools.nsc.doc.html.Page$.inlineToStr(Page.scala:18)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1(Page.scala:18)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1$adapted(Page.scala:18)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
        at scala.collection.immutable.List.flatMap(List.scala:352)
        at scala.tools.nsc.doc.html.Page$.inlineToStr(Page.scala:18)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1(Page.scala:18)
        at scala.tools.nsc.doc.html.Page$.$anonfun$inlineToStr$1$adapted(Page.scala:18)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
        at scala.collection.immutable.List.flatMap(List.scala:352)
        at scala.tools.nsc.doc.html.Page$.inlineToStr(Page.scala:18)
        at scala.tools.nsc.doc.html.page.IndexScript.$anonfun$shortDesc$2(IndexScript.scala:90)
        at scala.tools.nsc.doc.html.page.IndexScript.shortDesc(IndexScript.scala:89)
        at scala.tools.nsc.doc.html.page.IndexScript.$anonfun$packages$4(IndexScript.scala:47)
        at scala.tools.nsc.doc.html.page.IndexScript.$anonfun$packages$3(IndexScript.scala:41)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.collection.IterableLike.foreach(IterableLike.scala:71)
        at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike.map(TraversableLike.scala:234)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at scala.tools.nsc.doc.html.page.IndexScript.$anonfun$packages$1(IndexScript.scala:32)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
        at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
        at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
        at scala.collection.TraversableLike.map(TraversableLike.scala:234)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at scala.tools.nsc.doc.html.page.IndexScript.<init>(IndexScript.scala:28)
        at scala.tools.nsc.doc.html.page.IndexScript$.apply(IndexScript.scala:152)
        at scala.tools.nsc.doc.html.HtmlFactory.generate(HtmlFactory.scala:104)
        at scala.tools.nsc.doc.html.Doclet.generateImpl(Doclet.scala:23)
        at scala.tools.nsc.doc.doclet.Generator.generate(Generator.scala:23)
        at scala.tools.nsc.doc.DocFactory.generate$1(DocFactory.scala:128)
        at scala.tools.nsc.doc.DocFactory.document(DocFactory.scala:131)
        at xsbt.Runner.run(ScaladocInterface.scala:26)
        at xsbt.ScaladocInterface.run(ScaladocInterface.scala:10)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
        at sbt.compiler.AnalyzingCompiler.doc(AnalyzingCompiler.scala:73)
        at sbt.compiler.AnalyzingCompiler.doc(AnalyzingCompiler.scala:68)
        at sbt.Doc$$anonfun$scaladoc$1.apply(Doc.scala:23)
        at sbt.Doc$$anonfun$scaladoc$1.apply(Doc.scala:23)
        at sbt.RawCompileLike$$anonfun$prepare$1.apply(RawCompileLike.scala:64)
        at sbt.RawCompileLike$$anonfun$prepare$1.apply(RawCompileLike.scala:56)
        at sbt.RawCompileLike$$anonfun$cached$1$$anonfun$2$$anonfun$apply$1.apply(RawCompileLike.scala:49)
        at sbt.RawCompileLike$$anonfun$cached$1$$anonfun$2$$anonfun$apply$1.apply(RawCompileLike.scala:47)
        at sbt.Tracked$$anonfun$outputChanged$1.apply(Tracked.scala:84)
        at sbt.Tracked$$anonfun$outputChanged$1.apply(Tracked.scala:79)
        at sbt.RawCompileLike$$anonfun$cached$1.apply(RawCompileLike.scala:54)
        at sbt.RawCompileLike$$anonfun$cached$1.apply(RawCompileLike.scala:39)
        at sbt.Defaults$$anonfun$docTaskSettings$3.apply(Defaults.scala:856)
        at sbt.Defaults$$anonfun$docTaskSettings$3.apply(Defaults.scala:834)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
        at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:237)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.base/java.lang.Thread.run(Thread.java:844)
[error] (library/compile:doc) scala.reflect.internal.FatalError:
[error]   bad constant pool tag 20 at byte 108
[error]      while compiling: /localhome/doeraene/projects/scalajs/library/src/main/scala/scala/scalajs/reflect/annotation/EnableReflectiveInstantiation.scala
[error]         during phase: globalPhase=terminal, enteringPhase=parser
[error]      library version: version 2.12.3
[error]     compiler version: version 2.12.3
[error]   reconstructed args: -implicits -bootclasspath /home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar -feature -P:scalajs:sjsDefinedByDefault -P:scalajs:suppressExportDeprecations -unchecked -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar#http://www.scala-lang.org/api/2.12.3/ -doc-external-doc:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar#http://www.scala-js.org/api/scalajs-library/0.6.21-SNAPSHOT/ -classpath /home/doeraene/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.12.3.jar:/localhome/doeraene/projects/scalajs/library/target/scala-2.12/scalajs-library_2.12-0.6.21-SNAPSHOT.jar -encoding utf8 -deprecation -groups -Xfatal-warnings -d /localhome/doeraene/projects/scalajs/library/target/scala-2.12/api -Xplugin:/localhome/doeraene/projects/scalajs/compiler/target/scala-2.12/scalajs-compiler_2.12.3-0.6.21-SNAPSHOT.jar
[error]
[error]   last tree to typer: Ident(<argument>)
[error]        tree position: <unknown>
[error]             tree tpe: scala.scalajs.reflect.InstantiatableClass
[error]               symbol: <none>
[error]    symbol definition: <none> (a NoSymbol)
[error]       symbol package: <none>
[error]        symbol owners:
[error]            call site: package <root> in <none>
[error]
[error] <Cannot read source file>
[error] Total time: 7 s, completed Oct 10, 2017, 10:12:42 PM
```

File: javalib/src/main/scala/java/nio/ByteBuffer.scala [hunk 1]
```diff
@@ -83,6 +83,27 @@ abstract class ByteBuffer private[nio] (
...
+
+  @inline override def clear(): ByteBuffer =
+    super.clear().asInstanceOf[ByteBuffer]
+
+  @inline override def flip(): ByteBuffer =
+    super.flip().asInstanceOf[ByteBuffer]
+
+  @inline override def rewind(): ByteBuffer =
+    super.rewind().asInstanceOf[ByteBuffer]
+
```
Comment (gzm0) [inline javalib/src/main/scala/java/nio/ByteBuffer.scala]: I'd rather you return `this` than do a cast TBH :-S

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BufferAdapter.scala [hunk 2]
```diff
@@ -6,8 +6,15 @@ sealed abstract class BufferAdapter[BT <: Buffer, ET] {
...
-  def duplicate(): BufferType
+  /* Some methods have a Chain suffix because they are declared as abstract in
+   * java.nio.Buffer since Java 9, but with a result type of `Buffer` instead
+   * of the more specific `BufferType`. We use the `Chain` variant to be able
+   * to chain their application with further operations on the specific
+   * `BufferType`.
+   */
+
+  def sliceChain(): BufferType
+  def duplicateChain(): BufferType
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BufferAdapter.scala]: I don't get this. `BufferAdapter` does not have any inheritance relation to `Buffer`.

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BufferAdapter.scala]: The methods of `BufferAdapter` are pimped on the abstract type `BufferType`, which itself is `<: Buffer`. So you have code like `myBuffer.slice()` with `myBuffer: BufferType` which used to resolve to `bufferTypeToAdapter(myBuffer).slice()` (with result type `BufferType`) but now they don't go through the implicit conversion anymore, which means their result type is just `Buffer`.

Comment (gzm0) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BufferAdapter.scala]: Ah, I see :P

**Comment by sjrd** [comment]
Well, I am at a loss for the `library/doc` issue, so I worked around it by disabling Scaladoc generation altogether for `library` on JDK 9. The issue is filed as #3152.

Otherwise, I managed to make it work for Scala 2.10 and 2.11 as well, and fixed a couple other issues.

@SethTisue I think it's ready for another trial on the community build.

**Comment by SethTisue** [comment]
> I think it's ready for another trial on the community build.

`[info] Project scala-js----------------: SUCCESS (project rebuilt ok)`

👍 (but note this only indicates success building scala-js itself, we're not building downstream projects yet, for now the JDK9 build is still tiny and only exists on my laptop)

**Comment by sjrd** [comment]
I don't think Scala.js has any downstream project in the community build. The community build does not build the JS variants of projects.

**Pull request merged**